app/lib/package/export_api_to_bucket.dart

```diff
@@ -0,0 +1,297 @@
+      }
+      // Wait for 1 minutes for sanity, before trying again.
+      await Future.delayed(Duration(minutes: 1));
+    }
+  }
+
+  /// Gets and uploads the package name completion data.
+  Future<void> uploadPkgNameCompletionData() async {
+    final bytes = await searchBackend.getPackageNameCompletitionDataJsonGz();
+    for (final objectName in _apiPkgNameCompletitionDataNames()) {
+      if (await _isSameContent(objectName, bytes)) {
+        continue;
+      }
+      await uploadWithRetry(
+        _bucket,
+        objectName,
+        bytes.length,
+        () => Stream.value(bytes),
+        metadata: ObjectMetadata(
+          contentType: 'application/json; charset="utf-8"',
+          contentEncoding: 'gzip',
+          cacheControl:
+              'public, max-age=${_pkgNameCompletitionDataMaxAge.inSeconds}',
+        ),
+      );
+    }
+  }
+
+  /// Note: there is no global locking here, the full scan should be called
+  /// only once every day, and it may be racing against the incremental
+  /// updates.
+  @visibleForTesting
+  Future<void> fullPkgScanAndUpload() async {
+    final pool = Pool(_concurrency);
+    final futures = <Future>[];
+    await for (final mp in dbService.query<ModeratedPackage>().run()) {
+      final f =
+          pool.withResource(() => _processPkgUpdated(mp.asPkgUpdatedEvent()));
+      futures.add(f);
+    }
+    await Future.wait(futures);
+    futures.clear();
+
+    await for (final package in dbService.query<Package>().run()) {
+      final f = pool
+          .withResource(() => _processPkgUpdated(package.asPkgUpdatedEvent()));
+      futures.add(f);
+    }
+    await Future.wait(futures);
+    await pool.close();
+  }
+
+  @visibleForTesting
+  Future<void> incrementalPkgScanAndUpload(
+    GlobalLockClaim claim, {
+    Duration sleepDuration = const Duration(minutes: 2),
+  }) async {
+    final pool = Pool(_concurrency);
+    // The claim will be released after a day, another process may
+    // start to upload the API JSONs from scratch again.
+    final workUntil = clock.now().add(Duration(days: 1));
+
+    // start monitoring with a window of 7 days lookback
+    var lastQueryStarted = clock.now().subtract(Duration(days: 7));
+    while (claim.valid) {
+      final now = clock.now().toUtc();
+      if (now.isAfter(workUntil)) {
+        break;
+      }
+
+      // clear old entries from last seen cache
+      _pkgLastUpdated.removeWhere((key, event) =>
+          now.difference(event.updated) > const Duration(hours: 1));
+
+      lastQueryStarted = now;
+      final futures = <Future>[];
+      final eventsSince = lastQueryStarted.subtract(Duration(minutes: 5));
+      await for (final event in _queryRecentPkgUpdatedEvents(eventsSince)) {
+        if (!claim.valid) {
+          break;
+        }
+        final f = pool.withResource(() async {
+          if (!claim.valid) {
+            return;
+          }
+          await _processPkgUpdated(event);
```

Comment:
I'd suggest moving the contents of `_processPkgUpdate` in-line here.

Then you don't need `_pkgLastUpdated` to be a member, it can just be a variable in the scope of this function. That makes it much easier to track, and in some weird world where `incrementalPkgScanAndUpload` is invoked twice, concurrently, those two invocations won't interfere with each other by virtue of messing with `_pkgLastUpdated`.

Reply:
We may move `_processPkgUpdated` inline, but we will need to update the (daily) full scan to repeat most of its logic. Which would be fine if we don't want to skip the update based on a timestamp... which is probably fine, as that process may not be triggered on the same machine at all.

Reply:
But the daily full scan should not interfere with `_pkgLastUpdated`, and certainly shouldn't populate it. That'll just burn an awful lot of memory for no purpose.

Especially, if it's not running on the instance that has the lock and is currently scanning.

I think my point was that `_pkgLastUpdated` should be private to `incrementalPkgScanAndUpload`. You can also just pull those bits out of `_processPkgUpdated`

Reply:
Agreed and done.